### PR TITLE
[69624] Scheduler API Support

### PR DESCRIPTION
**Pull request by mrashed-dev**
# Description
This PR enables SDK support for the [Scheduler API](https://developer.nylas.com/docs/api/scheduler/#overview).

# Usage
To create a new Scheduler page:
```python
scheduler = nylas.scheduler.create()
scheduler.access_tokens = ["ACCESS_TOKEN"]
scheduler.name = "Python SDK Example"
scheduler.slug = "py_example_1"
scheduler.save()
```

To return all Scheduler pages:
```python 
scheduler_list = nylas.scheduler.all()
```

To return a single Scheduler page:
```python
scheduler = nylas.scheduler.get('SCHEDULER_ID')
```

To update a Scheduler page:
```python
scheduler = nylas.scheduler.get('SCHEDULER_ID')
scheduler.name = "Updated page name"
scheduler.save()
```

To delete a Scheduler page:
```python
nylas.scheduler.delete('SCHEDULER_ID')
```

To get available calendars for a Scheduler page:
```python
scheduler = nylas.scheduler.get('SCHEDULER_ID')
calendars = scheduler.get_available_calendars()
```

To upload an image:
```python
scheduler = nylas.scheduler.get('SCHEDULER_ID')
scheduler.upload_image(content_type: "image/png", object_name: "test.png")
```

Checking Provider Availability
```python
# Google Availability
google_availability = nylas.scheduler.get_google_availability()

# Office 365 Availability
o365_availability = nylas.scheduler.get_office_365_availability()
```

Get page information/configuration
```python
page_config = nylas.scheduler.get_page_slug('slug')
```

Retrieve available time slots
```python
available_timeslots = nylas.scheduler.get_available_time_slots('slug')
```

Book a time slot
```python
slot = SchedulerTimeSlot.create(nylas)
slot.account_id = "test-account-id"
slot.calendar_id = "test-calendar-id"
slot.emails = ["recipient@example.com"]
slot.host_name = "Host"
slot.start = datetime.utcfromtimestamp(1636728347)
slot.end = datetime.utcfromtimestamp(1636731958)

timeslot_to_book = SchedulerBookingRequest.create(nylas)
timeslot_to_book.additional_values = {
    "test": "yes",
}
timeslot_to_book.email = "recipient@example.com"
timeslot_to_book.locale = "en_US"
timeslot_to_book.name = "Recipient Doe"
timeslot_to_book.timezone = "America/New_York"
timeslot_to_book.slot = slot

booking_confirmation = nylas.scheduler.book_time_slot("slug", timeslot_to_book)
```

Confirm a booking
```python
booking_confirmation = nylas.scheduler.confirm_booking('slug', 'edit-hash');
```

Cancel a booking
```python
nylas.scheduler.cancel_booking('slug', 'edit-hash', 'reason');
```

# License
I confirm that this contribution is made under the terms of the MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: nylas/client/restful_models.py [hunk 1]
```diff
@@ -688,6 +688,49 @@ def __init__(self, api):
...
+        response_body = response.json()
+        for body in response_body:
+            for i in range(len(body["calendars"])):
+                body["calendars"][i] = Calendar.create(self.api, **body["calendars"][i])
+
+        return response_body
+
+    def upload_image(self, content_type, object_name):
+        if not self.id:
+            raise ValueError("Cannot upload an image to a page without an ID")
```
[the remark this answers is not in the document]
Comment (mrashed-dev) [reply nylas/client/restful_models.py:743]: Yes, those are added in now!

File: nylas/client/client.py [hunk 2]
```diff
@@ -551,7 +567,11 @@ def _update_resource(self, cls, id, data, **kwargs):
         converted_data = create_request_body(data, cls.datetime_attrs)
         response = session.put(url, json=converted_data)
 
-        result = _validate(response).json()
+        result = _validate(response)
+        return result.json()
```
[the remark this answers is not in the document]
Comment (mrashed-dev) [reply nylas/client/client.py:614]: Despite what the method name suggests, it returns a dictionary object not a JSON string!

**Comment by mrashed-dev** [comment]
@tbadun Added the missing functionality! This PR should cover all of the Scheduler API endpoints.

File: nylas/client/client.py [hunk 3]
```diff
@@ -466,11 +474,13 @@ def _get_resources(self, cls, extra=None, **filters):
...
     def _get_resource_raw(
-        self, cls, id, extra=None, headers=None, stream=False, **filters
+        self, cls, id, extra=None, headers=None, stream=False, path=None, **filters
     ):
         """Get an individual REST resource"""
+        if path is None:
+            path = cls.collection_name
         postfix = "/{}".format(extra) if extra else ""
-        path = "/{}".format(cls.collection_name) if cls.collection_name else ""
+        path = "/{}".format(path) if path else ""
```
Comment (tbadun) [inline nylas/client/client.py:483]: what would a false-y `path` look like here if you're setting it at 481?

Comment (mrashed-dev) [reply nylas/client/client.py:483]: If `path` isn't passed in, path is set to `cls.collection_name`. This is just on the off chance that the `collection_name` isn't set for the collection passed in.

**Review by tbadun — APPROVED** [review]
👍  (barring my feelings about mock 🙃)

**Pull request merged**